Expand case conversion modes for slugified content

elbeardmorez opened the pull request:
For your consideration. Regurgitating the single commit:
```
- Facilitates additional 'upper' and 'camel' case conversion modes when slugification is enabled
- Defaulting to mode 'lower' in order to maintain backward compatibility
- Removes the existing '-u' / '--keep-upper' option as its functionality is entirely consumed by this new option. The direct replacement being '-x none' / '--case-convert=none'
```
All modes appear to work as expected. modifying the mozilla _slugify_ package was never going to happen (this end) so the only real change is the couple of lines in the `slugify_preset()` and the coaxing / massaging of the existing -> new args. My only qualm is  how it looks - 'slighty nasty' - I wonder on your tolerance on cmdline opt cleanliness, yes it's 'pythonic' (I assume given the stdlib parser) but 'man does it look ugly':

```
  -k, --keep-spaces     Retain whitespace in filenames
  -x {lower,upper,camel,none}, --case-convert {lower,upper,camel,none}
                        Specify the char case conversion logic, default: lower
  --no-confirm          Override confirmation prompts. Use with caution
  --embed-genres        Embed album/track genres
```
ps. having grepped for instances of `--keep-upper` that I needed to replace, I noted that the README.rst has no reference to the last two (assuming newest) options(?!) `--no-confirm` / `--embed-genres`, perhaps they should be added for completeness

elbeardmorez committed:
Options, expand case conversion modes

- Facilitates additional 'upper' and 'camel' case conversion modes
  when slugification is enabled
- Defaulting to mode 'lower' in order to maintain backward
  compatibility
- Removes the existing '-u' / '--keep-upper' option as its
  functionality is entirely consumed by this new option. The direct
  replacement being '-x none' / '--case-convert=none'

elbeardmorez commented:
Force pushed - apols. I made a last minute option name change `--case-mode` -> `--case-convert` before the original submission to 'help' the user understand the option's purpose, and that wasn't tested with a clean profile (deleted generated json). Hence I missed the need for the `dest` parameter.

 `--case-convert` seems appropriate for the user to see in the 'help' output given the default isn't 'none' (but lower, as per the original status quo), but `case_mode` is used throughout the code as I feel it's a better fit. FWIW I personally think the default should be 'none' as opposed to 'lower', but that's not my call.

Evolution0 commented:
Yes the documentation is a bit out of date as I've mainly popped in to fix broken things, a pass of the docs and other things needs to be done soon.

I'm fine with the PR as is as long as it doesn't break people's configs

elbeardmorez committed:
Config, make update 'private'

- Nothing external uses the 'update_with_dict()', so keep the pattern
  of '_'-prefix consistent

elbeardmorez committed:
Config, fix comment typo

elbeardmorez committed:
Config, add option migration pattern

- Implements reverse migration of a future option. Arbitrary string
  values are used in place of 'as yet unknown' constants

elbeardmorez commented:
Right, good call! I've added a few tweaks to the config initialisation / processing for migrations.

Switching back to pre-feature and running: my resulting config looked like this:
```
{"base-dir": "/root", "template": "%{artist}/%{album}/%{track} - %{title}", "overwrite": false, "no-art": false, "embed-art": false, "embed-lyrics": false, "group": false, "no-slugify": false, "ok-chars": "-_~", "space-char": "-", "ascii-only": false, "keep-spaces": false, "case-mode": "lower", "no-confirm": false, "debug": false, "embed-genres": false}
```
Note the lack of existing 'keep_upper' option and the retention of the invalid / unknown 'case_mode' option. We'd need to be dropping 'case_mode' and adding back the 'keep_upper' in this (non-conventional 'backward') instance.

The existing code silently ignored unknown options, and also, would only dump a config instance (consisting of defaults) if *no* existing config was present. Ignoring unknown options means that any deprecated options, or manually input garbage for that matter, would persist in the config file until it was deleted. Further, only dumping a copy of the config instance if no existing config file was present meant that any new config options that came down the line would never have their defaults make their way into the persisted config file.

Post tweaks, these shortfalls are remedied. Unknown options are sent to `_migrate_option()` where they are migrated (forward or backward) where supported. and when unsupported, dropped. The expectation is that the given pattern is followed for any further migrations. There is more 'noise' than before on encountering unknown config options and I think this is rightly so.

The instance whereby config file options are either manually removed (as unlikely as this might be) or new options are added (far more likely) is dealt with via set comparison and re-persistence of config file on any diff.

So in testing, vacillating between HEAD~1 <-> HEAD with invalid configs I see stderr output like:
```
Reverse migration of config option: 'case_mode=none' -> 'keep_upper=True'
Modified configuration has been written to `path/to/.config/bandcamp-dl.json'.
```
and, having added garbage and returned to HEAD..
```
Forward migration of config option: 'keep_upper=True' -> 'case_mode=none'
Dropping unknown config option 'sponge_bob=square-pants'
Modified configuration has been written to `path/to/.config/bandcamp-dl.json'.
```
and then having manually set my preference and back at HEAD~1
```
Warning, lossy reverse migration, new value 'camel' is not backwards compatible
Reverse migration of config option: 'case_mode=camel' -> 'keep_upper=True'
Modified configuration has been written to `path/to/.config/bandcamp-dl.json'.
```
because the expanded `--case-convert` option has more functionality (and thus option values) than that which it replaces (`--keep-upper`). etc. etc.. Should be good to go now, thanks again for the prompt.

Evolution0 commented:
Perfect, merging.

Evolution0 commented:
I'll push a build sometime soon, currently setting up my mobile workstation (which will also allow me to start working on things like a flake and dockerfile)
